pageserver/src/object_repository.rs

```diff
@@ -987,6 +997,45 @@ const fn timeline_metadata_key(timelineid: ZTimelineId) -> ObjectKey {
     }
 }
 
+///
+/// Helper functions to deserialize ObjectValue, when the caller knows what kind of
+/// a value it should be.
+///
+/// There are no matching helper functions for serializing. Instead, there are
+/// `put_page_entry`, `put_relsize_entry`, and `put_timeline_metadata_entry` helper
+/// functions in ObjectTimeline that both construct the right kind of key and
+/// serialize the value in the same call.
+///
+impl ObjectValue {
+    fn des_page(v: &[u8]) -> Result<PageEntry> {
+        match ObjectValue::des(&v)? {
+            ObjectValue::Page(p) => Ok(p),
+            _ => {
+                bail!("Invalid object kind, expected a page entry");
+            }
+        }
+    }
+
+    fn des_relsize(v: &[u8]) -> Result<RelationSizeEntry> {
+        match ObjectValue::des(&v)? {
+            ObjectValue::RelationSize(rs) => Ok(rs),
+            _ => {
+                bail!("Invalid object kind, expected a relation size entry");
+            }
+        }
+    }
+
+    fn des_timeline_metadata(v: &[u8]) -> Result<MetadataEntry> {
+        match ObjectValue::des(&v)? {
+            ObjectValue::TimelineMetadata(t) => Ok(t),
+            _ => {
+                bail!("Invalid object kind, expected a timeline metadata entry");
+            }
+        }
+    }
```

Comment:
I like having these helpers, the code can get pretty messy with `if let`/`match` inside functions.

It'd be cool if we could associate the matching key/value variants with Rust types, making it impossible for programmers to mismatch the key and value types

Reply:
> It'd be cool if we could associate the matching key/value variants with Rust types, making it impossible for programmers to mismatch the key and value types

Yeah. I suppose we could introduce a "KeyValuePair" enum that would have different variants for the valid key and value combinations. But that'd be a bigger refactoring and I'm not sure if the end result would really be better in the end.